libs/sdk-core/src/breez_services.rs

```diff
@@ -2096,56 +2078,68 @@ impl Receiver for PaymentReceiver {
                         cltv_expiry_delta: lsp_info.time_lock_delta as u64,
                         htlc_minimum_msat: Some(lsp_info.min_htlc_msat as u64),
                         htlc_maximum_msat: None,
-                    };
-
-                    info!("Adding LSP hop as routing hint: {:?}", lsp_hop);
-                    lsp_hint = Some(RouteHint {
-                        hops: vec![lsp_hop],
-                    });
-                }
-                None => info!("No available channel ID for route hint"),
+                    }],
+                };
+                info!("Adding open channel hint: {:?}", open_channel_hint);
+                Some(add_routing_hints(
+                    invoice.clone(),
+                    false,
+                    &vec![open_channel_hint],
+                    req.amount_msat,
+                )?)
             }
-        }
-
-        // We only create a new invoice if we need to add the lsp hint or change the amount
-        if lsp_hint.is_some() || req.amount_msat != destination_invoice_amount_msat {
-            // create the large amount invoice
-            let raw_invoice_with_hint = add_lsp_routing_hints(
-                invoice.clone(),
-                !open_channel_needed,
-                lsp_hint,
-                req.amount_msat,
-            )?;
-
-            info!("Routing hint added");
-            let signed_invoice_with_hint = self.node_api.sign_invoice(raw_invoice_with_hint)?;
-            info!("Signed invoice with hint = {}", signed_invoice_with_hint);
+            // In case we don't need to open a channel and we have a channel with our lsp then we only ensure it
+            // exists as part of the invoice routing hints (merging).
+            (false, Some(h), _) => {
+                // In case we need to open a channel and the invoice has no routing hints we replace them with ours.
+                info!("Adding lsp hint: {:?}", h);
+                Some(add_routing_hints(
+                    invoice.clone(),
+                    true,
+                    &vec![h],
+                    req.amount_msat,
+                )?)
+            }
+            // In case we don't need to open a channel and the invoice has no routing hints we replace them with ours.
+            (false, None, true) => {
+                info!("Adding custom hints: {:?}", routing_hints);
+                Some(add_routing_hints(
+                    invoice.clone(),
+                    false,
+                    &routing_hints,
+                    req.amount_msat,
+                )?)
+            }
+            (_, _, _) => None,
+        };
 
-            parsed_invoice = parse_invoice(&signed_invoice_with_hint)?;
+        if let Some(raw_invoice) = optional_modified_invoice {
+            let signed_invoice = self.node_api.sign_invoice(raw_invoice)?;
+            info!("Signed invoice with hint = {}", signed_invoice);
+            parsed_invoice = parse_invoice(&signed_invoice)?;
```

Comment:
I think for the normal case (not opening channel, have channel with lsp, have GL lsp route hint), this will still result in 2 invoices every time

Reply:
Good catch. Fixed